Pass args '--upgrade-branch' for Cookiecutter template git branch name to run upgrade from

yadavnikhil opened the pull request:
I need logic to be able to pass branch name for git repo of cookiecutter template when upgrading projects. 
I've added an additional argument that can be passed to cli and to cookiecutter checkout param. Default will be 'master' - the default for cookiecutter.
As default branch will be master, it will not impact existing functionality.

ddelange reviewed (commented):
Hi @yadavnikhil, thanks for your contribution. Please check out Travis for failing checks.

One suggestion: it would be better if the default value for this flag [remains](https://github.com/cookiecutter/cookiecutter/blob/1.7.2/cookiecutter/main.py#L27) None, instead of a hardcoded `master`, especially now that new repositories get created with `main` as default branch unless specified otherwise (ref https://github.com/github/renaming).

ddelange commented:
> Please check out Travis for failing checks.

Seems that mypy started complaining about the dummy cookiecutter template in the tests folder. Limiting mypy to the actual test files should make that check pass:

```console
$ mypy cookiecutter_project_upgrader tests/*.py
Success: no issues found in 8 source files
```

yadavnikhil commented:
Yes you right, None will be also fine, I will change the default value to None. Yes mypy is giving errors . Will try to change command as mentioned in tox ini and check if it passes

yadavnikhil commented:
Unfortunately it looks like tox.ini doesn't still support wildcard files in its configurations, though I see some work going on in that area to support it,
Meanwhile I modified tox.ini to include all py files from tests folder instead of '*.py' to make it pass.
We need to change it back to wildcard once tox support expanding paths.

yadavnikhil commented:
Added one more commit to pass a --zip-file argument to use an zip file path/url of templates instead of git repo. 
For coverage & tests, I've added tests for local zip files only. For remote zip files, we rely validity check from cookiecutter package.

ddelange commented on cookiecutter_project_upgrader/cli.py:
```diff
@@ -11,6 +13,9 @@
               default="docs/cookiecutter_input.json", help="Default: docs/cookiecutter_input.json")
 @click.option('--branch', '-b', default="cookiecutter-template", help="Default: cookiecutter-template")
+@click.option('--upgrade-branch', '-u',
+              help="Cookiecutter template git branch name to upgrade from?")
```
```suggestion
              help="Optional branch name of cookiecutter template to checkout before upgrading.")
```
to avoid questions in helptexts

ddelange commented on cookiecutter_project_upgrader/cli.py:
```diff
@@ -20,16 +25,20 @@
               help='Git pathspecs to exclude from the update commit, e.g. -e "*.py" -e "tests/".')
-def main(context_file: str, branch: str,
-         interactive: bool, merge_now: Optional[bool],
+def main(context_file: str, branch: str, upgrade_branch: str,
```
```suggestion
def main(context_file: str, branch: str, upgrade_branch: Optional[str],
```

ddelange commented on cookiecutter_project_upgrader/cli.py:
```diff
@@ -45,3 +54,29 @@ def _load_context(context_file: str):
+
+def _is_valid_file(zip_file: str):
+    # we check only for local files for valid, for remote URLs, cookiecutter can do the test if its valid.
+    REPO_REGEX = re.compile(
+        r"""
+        ((((git|hg)\+)?(git|ssh|file|https?):(//)?)
+        |
+        (\w+@[\w\.]+)
+        )
+        """,
+        re.VERBOSE,
+        )
+    try:
+        if not bool(REPO_REGEX.match(zip_file)):
+            if ZipFile(Path(zip_file)).testzip() is None:
+                return True
```
```suggestion
    zip_path = Path(zip_file)
    try:
        if zip_path.exists():
            ZipFile(zip_path).testzip()
```

I believe this will have the same behaviour (always True, unless `testzip` throws an error), and avoids having a regexp to maintain :)

I don't think any of the formats from the regexp would actually fail to parse into a Path: 
```pycon
>>> Path('git+https://github.com').exists()
False
```

yadavnikhil replied:
Yes sure, using above we don't need to except FileNotFound error.

ddelange commented on cookiecutter_project_upgrader/cli.py:
```diff
@@ -45,3 +54,29 @@ def _load_context(context_file: str):
+        if not bool(REPO_REGEX.match(zip_file)):
+            if ZipFile(Path(zip_file)).testzip() is None:
+                return True
+    except BadZipFile:
+        click.echo(f"Bad Zip file at: {zip_file}\n"
+                   f"Make sure your zip file is correct.")
+        exit(1)
+    except FileNotFoundError:
+        click.echo(f"Zip file not found at: {zip_file}\n"
+                   f"Make sure your path is correct.")
+        exit(1)
```
nit: click [supports](https://click.palletsprojects.com/en/7.x/api/#click.echo) echoing to stderr with `err=True`

also realised that with my suggestion above, this error won't be raised, for which we can depend on the upstream error from cookiecutter itself:

```suggestion
    except BadZipFile:
        click.echo(f"Bad Zip file at: {zip_file}\n"
                   f"Make sure your zip file is correct.", err=True)
        exit(1)
```

ddelange commented on tests/test_logic.py:
```diff
@@ -147,6 +150,47 @@ def test_change_project_slug(cookiecutter_template_directory: Path,
+                                             "README.rst").write_text("updated readme for zip test")
+    subprocess.run(["git", "add", "-A"], cwd=str(cookiecutter_template_directory), check=True)
+    subprocess.run(["git", "commit", "-m", "updated readme for zip test"],
+                   cwd=str(cookiecutter_template_directory), check=True)
```
nice work with the test! as another nit, I can offer you shorthands `subprocess.check_call` which does the above, and `subprocess.check_output` to additionally return the output of the command that was run

yadavnikhil replied:
I think subprocess.run() will be more appropriate here which was added in python 3.5 and is recommended way. Since we running tests with python v3.5+ we should be fine with run() rather than older call. What do you think?

ddelange commented on cookiecutter_project_upgrader/logic.py:
```diff
@@ -46,7 +46,8 @@ def _git_repository_has_local_changes(git_repository: Path):
 
 def update_project_template_branch(context: MutableMapping[str, str], project_directory: str, branch: str,
-                                   merge_now: Optional[bool], push_template_branch_changes: Optional[bool],
+                                   upgrade_branch: str, merge_now: Optional[bool],
```
```suggestion
                                   upgrade_branch: Optional[str], merge_now: Optional[bool],
```

ddelange reviewed (commented):
Just a few suggestions :)

yadavnikhil commented:
Looks like travis-ci - 'xenial' distribution is failing for py36 tests for rust compilers(Failed build link: [here](https://travis-ci.org/github/thomasjahoda/cookiecutter_project_upgrader/builds/758018715)).
To test I changed to 'bionic' distribution & it worked fine.

ddelange reviewed (approved):
Nothing more from my side :) will leave it open another day in case @thomasjahoda wants to have a look.

For the failing build on xenial: `cryptography` wheels need a recent pip to unpack them, and it seems xenial came with pip10 (logs L552) instead of pip21. Since bionic is newer it probably came with a new enough version of pip by default. let's keep it!

ddelange commented:
Hi @yadavnikhil, just as update: Travis keeps [crashing](https://travis-ci.org/github/thomasjahoda/cookiecutter_project_upgrader/jobs/758220283) without traceback no matter how often I rerun the deploy stage.

Planning a migration to GitHub Actions, hopefully before next week.

yadavnikhil commented:
@ddelange @thomasjahoda I'm waiting for new version to be available in PYPI. Can we have it soon?
Thanks,

ddelange commented:
Hi @yadavnikhil, thanks for the reminder. For now, I've manually pushed v1.3.0 to PyPI, whilst awaiting #4.
